retry some failed record changes

britneywright opened the pull request:
close #842 

Retry failed record changes if the record change error is Refused during the apply step of the finite state machine, or if the record change error is Try Again during the verify step of the finiite state machine in the RecordSetChangeHandler. Raising an error in the RecordSetChangeHandler causes the change to go back on the MySqlMessageQueue.

Tested against a commonly failing scenario, changing an A record to a CNAME through a batch change. Currently the A record deletes successfully but the CNAME fails in the backend. With the re-queue the CNAME add appears to succeed.

record statuses based on current flow:
<img width="1009" alt="Screen Shot 2019-11-20 at 9 12 29 PM" src="https://user-images.githubusercontent.com/4439228/69345166-b016a080-0c3e-11ea-91ba-73dd3456c9ba.png">

record statuses based on updated flow:
<img width="1005" alt="Screen Shot 2019-11-20 at 9 12 46 PM" src="https://user-images.githubusercontent.com/4439228/69345178-b4db5480-0c3e-11ea-9c6f-eb9d79f7cf02.png">



Need to add/update more tests.

mitruly commented on modules/api/src/main/scala/vinyldns/api/engine/RecordSetChangeHandler.scala:
```diff
@@ -225,32 +236,27 @@ object RecordSetChangeHandler {
-              )
+  /* Step 3: Verify the record was created. */
+  private def verify(change: RecordSetChange, dnsConn: DnsConnection): IO[ProcessorState] =
+    getProcessingStatus(change, dnsConn).flatMap {
```
We don't need to `flatMap` with the removal of nested `loop`; this can be consolidated to:
```
    getProcessingStatus(change, dnsConn).map {
      case AlreadyApplied(_) => Completed(change.successful)
      case ReadyToApply(_) | Retry(_) => Retrying(change)
      case Failure(_, message) =>
        Completed(
          change.failed(
            s"Failed verifying update to DNS for change ${change.id}:${change.recordSet.name}: $message"
          )
        )
    }
```

britneywright commented:
I added to the existing batch change func test that tests updating records. Made an A record then tested deleting that A record and adding a CNAME with the same name. Before the RecordSetChangeHandler updates the result was a PartialFailure, after the updates it was Complete. I did have to increase the MAX_RETRIES in the func tests. I changed from 40 to 100 arbitrarily, but i'll test to see how low it can be and pass consistently.

pauljamescleary commented on modules/api/functional_test/vinyldns_python.py:
```diff
@@ -26,7 +26,7 @@
 __all__ = [u'VinylDNSClient', u'MAX_RETRIES', u'RETRY_WAIT']
 
-MAX_RETRIES = 40
+MAX_RETRIES = 100
```
oof, I am going to go out on a limb here and say this is something we do not want to e2e test, as we just spent a lot of time speeding up our builds.  Perhaps just leave it to unit tests down below where we have more control and tests are fast

britneywright replied:
For what it's worth the run time seems to be about the same with this change.

mitruly replied:
Think having a test should be fine @pauljamescleary since in our SQS requeue, we set to 10 seconds (instead of defaulting to default visibility timeout). That said, having this as a separate test for changing a record type in our functional tests would probably be better (1. so we don't have to increase `MAX_RETRIES` and 2. so the test can run in parallel with others).

britneywright commented:
I'll move the update record type to a separate func test and see if that allows me to bring down the retries.

britneywright commented:
The func tests passed once with the MAX_RETRIES back at 40, but then i rebased and they continually failed. Also failed at 60. Passed at 75 once. will rerun them couple times to confirm it's consistently passing.
